packages/x-date-pickers/src/PickersLayout/PickersLayout.types.ts

```diff
@@ -43,6 +43,10 @@ export interface ExportedPickersLayoutSlotsComponentsProps<TValue, TView extends
    * Props passed down to the layoutRoot component.
    */
   layout?: Partial<PickersLayoutProps<TValue, TView>>;
+  /**
+   * Props passed down to the toolbar component.
+   */
+  toolbar?: ExportedBaseToolbarProps;
```

Comment:
Most of the time (if not always), we are using `ExportedXXX` to say that an interface contains only props exposed on the closest public component (here the picker).

The main goal being to avoid using `Omit` on the component typing to keep it clean and avoid inconsistencies between pickers.

If I understand correctly, your problem is that you need the slot `toolbar` to be accessible at the headless level (`useDesktopPicker` and `useMobilePicker`.
But if you just move it to the `Exposed` slots of the layout, then the pickers which are adding other props to the `toolbar` slots complains.

The solution you took is to `Omit` the `toolbar` slot coming from the headless interfaces.
It works, but if we could avoid introducing back any `Omit` at the component level it would be great.

An alternative is to create a `Exported` interface for the slots at the headless level:

```diff
--- a/packages/x-date-pickers/src/DesktopDatePicker/DesktopDatePicker.types.ts
+++ b/packages/x-date-pickers/src/DesktopDatePicker/DesktopDatePicker.types.ts
@@ -1,6 +1,6 @@
 import {
   UseDesktopPickerSlotsComponent,
-  UseDesktopPickerSlotsComponentsProps,
+  ExportedUseDesktopPickerSlotsComponentsProps,
   DesktopOnlyPickerProps,
 } from '../internals/hooks/useDesktopPicker';
 import {
@@ -19,7 +19,7 @@ export interface DesktopDatePickerSlotsComponent<TDate>
 
 export interface DesktopDatePickerSlotsComponentsProps<TDate>
   extends BaseDatePickerSlotsComponentsProps<TDate>,
-    UseDesktopPickerSlotsComponentsProps<TDate, DateView> {}
+    ExportedUseDesktopPickerSlotsComponentsProps<TDate, DateView> {}
 
 export interface DesktopDatePickerProps<TDate>
   extends BaseDatePickerProps<TDate>,
diff --git a/packages/x-date-pickers/src/DesktopDateTimePicker/DesktopDateTimePicker.types.ts b/packages/x-date-pickers/src/DesktopDateTimePicker/DesktopDateTimePicker.types.ts
index 13b04c563..979c499cb 100644
--- a/packages/x-date-pickers/src/DesktopDateTimePicker/DesktopDateTimePicker.types.ts
+++ b/packages/x-date-pickers/src/DesktopDateTimePicker/DesktopDateTimePicker.types.ts
@@ -1,6 +1,6 @@
 import {
   UseDesktopPickerSlotsComponent,
-  UseDesktopPickerSlotsComponentsProps,
+  ExportedUseDesktopPickerSlotsComponentsProps,
   DesktopOnlyPickerProps,
 } from '../internals/hooks/useDesktopPicker';
 import {
@@ -22,7 +22,7 @@ export interface DesktopDateTimePickerSlotsComponent<TDate>
 
 export interface DesktopDateTimePickerSlotsComponentsProps<TDate>
   extends BaseDateTimePickerSlotsComponentsProps<TDate>,
-    Omit<UseDesktopPickerSlotsComponentsProps<TDate, DateOrTimeView>, 'toolbar'> {}
+    ExportedUseDesktopPickerSlotsComponentsProps<TDate, DateOrTimeView> {}
 
 export interface DesktopDateTimePickerProps<TDate>
   extends BaseDateTimePickerProps<TDate>,
diff --git a/packages/x-date-pickers/src/DesktopTimePicker/DesktopTimePicker.types.ts b/packages/x-date-pickers/src/DesktopTimePicker/DesktopTimePicker.types.ts
index 51e6e1293..994d41755 100644
--- a/packages/x-date-pickers/src/DesktopTimePicker/DesktopTimePicker.types.ts
+++ b/packages/x-date-pickers/src/DesktopTimePicker/DesktopTimePicker.types.ts
@@ -1,6 +1,6 @@
 import {
   UseDesktopPickerSlotsComponent,
-  UseDesktopPickerSlotsComponentsProps,
+  ExportedUseDesktopPickerSlotsComponentsProps,
   DesktopOnlyPickerProps,
 } from '../internals/hooks/useDesktopPicker';
 import {
@@ -19,7 +19,7 @@ export interface DesktopTimePickerSlotsComponent<TDate>
 
 export interface DesktopTimePickerSlotsComponentsProps<TDate>
   extends BaseTimePickerSlotsComponentsProps,
-    Omit<UseDesktopPickerSlotsComponentsProps<TDate, TimeView>, 'toolbar'> {}
+    ExportedUseDesktopPickerSlotsComponentsProps<TDate, TimeView> {}
 
 export interface DesktopTimePickerProps<TDate>
   extends BaseTimePickerProps<TDate>,
diff --git a/packages/x-date-pickers/src/MobileDatePicker/MobileDatePicker.types.ts b/packages/x-date-pickers/src/MobileDatePicker/MobileDatePicker.types.ts
index 99e1576ba..83639c495 100644
--- a/packages/x-date-pickers/src/MobileDatePicker/MobileDatePicker.types.ts
+++ b/packages/x-date-pickers/src/MobileDatePicker/MobileDatePicker.types.ts
@@ -1,7 +1,7 @@
 import {
   UseMobilePickerSlotsComponent,
-  UseMobilePickerSlotsComponentsProps,
   MobileOnlyPickerProps,
+  ExportedUseMobilePickerSlotsComponentsProps,
 } from '../internals/hooks/useMobilePicker';
 import {
   BaseDatePickerProps,
@@ -19,7 +19,7 @@ export interface MobileDatePickerSlotsComponent<TDate>
 
 export interface MobileDatePickerSlotsComponentsProps<TDate>
   extends BaseDatePickerSlotsComponentsProps<TDate>,
-    UseMobilePickerSlotsComponentsProps<TDate, DateView> {}
+    ExportedUseMobilePickerSlotsComponentsProps<TDate, DateView> {}
 
 export interface MobileDatePickerProps<TDate>
   extends BaseDatePickerProps<TDate>,
diff --git a/packages/x-date-pickers/src/MobileDateTimePicker/MobileDateTimePicker.types.ts b/packages/x-date-pickers/src/MobileDateTimePicker/MobileDateTimePicker.types.ts
index 9abfefceb..ea2247309 100644
--- a/packages/x-date-pickers/src/MobileDateTimePicker/MobileDateTimePicker.types.ts
+++ b/packages/x-date-pickers/src/MobileDateTimePicker/MobileDateTimePicker.types.ts
@@ -1,6 +1,6 @@
 import {
   UseMobilePickerSlotsComponent,
-  UseMobilePickerSlotsComponentsProps,
+  ExportedUseMobilePickerSlotsComponentsProps,
   MobileOnlyPickerProps,
 } from '../internals/hooks/useMobilePicker';
 import {
@@ -19,7 +19,7 @@ export interface MobileDateTimePickerSlotsComponent<TDate>
 
 export interface MobileDateTimePickerSlotsComponentsProps<TDate>
   extends BaseDateTimePickerSlotsComponentsProps<TDate>,
-    Omit<UseMobilePickerSlotsComponentsProps<TDate, DateOrTimeView>, 'toolbar'> {}
+    ExportedUseMobilePickerSlotsComponentsProps<TDate, DateOrTimeView> {}
 
 export interface MobileDateTimePickerProps<TDate>
   extends BaseDateTimePickerProps<TDate>,
diff --git a/packages/x-date-pickers/src/MobileTimePicker/MobileTimePicker.types.ts b/packages/x-date-pickers/src/MobileTimePicker/MobileTimePicker.types.ts
index 4f03f5962..8dacf5a07 100644
--- a/packages/x-date-pickers/src/MobileTimePicker/MobileTimePicker.types.ts
+++ b/packages/x-date-pickers/src/MobileTimePicker/MobileTimePicker.types.ts
@@ -1,6 +1,6 @@
 import {
   UseMobilePickerSlotsComponent,
-  UseMobilePickerSlotsComponentsProps,
+  ExportedUseMobilePickerSlotsComponentsProps,
   MobileOnlyPickerProps,
 } from '../internals/hooks/useMobilePicker';
 import {
@@ -19,7 +19,7 @@ export interface MobileTimePickerSlotsComponent<TDate>
 
 export interface MobileTimePickerSlotsComponentsProps<TDate>
   extends BaseTimePickerSlotsComponentsProps,
-    Omit<UseMobilePickerSlotsComponentsProps<TDate, TimeView>, 'toolbar'> {}
+    ExportedUseMobilePickerSlotsComponentsProps<TDate, TimeView> {}
 
 export interface MobileTimePickerProps<TDate>
   extends BaseTimePickerProps<TDate>,
diff --git a/packages/x-date-pickers/src/PickersLayout/PickersLayout.types.ts b/packages/x-date-pickers/src/PickersLayout/PickersLayout.types.ts
index e0b2e19ee..7908f61b5 100644
--- a/packages/x-date-pickers/src/PickersLayout/PickersLayout.types.ts
+++ b/packages/x-date-pickers/src/PickersLayout/PickersLayout.types.ts
@@ -43,10 +43,6 @@ export interface ExportedPickersLayoutSlotsComponentsProps<TValue, TView extends
    * Props passed down to the layoutRoot component.
    */
   layout?: Partial<PickersLayoutProps<TValue, TView>>;
-  /**
-   * Props passed down to the toolbar component.
-   */
-  toolbar?: ExportedBaseToolbarProps;
 }
 
 export interface PickersLayoutSlotsComponent<TValue, TView extends DateOrTimeView>
@@ -68,6 +64,10 @@ export interface PickersLayoutSlotsComponentsProps<TValue, TView extends DateOrT
    * Props passed down to the tabs component.
    */
   tabs?: ExportedBaseTabsProps;
+  /**
+   * Props passed down to the toolbar component.
+   */
+  toolbar?: ExportedBaseToolbarProps;
 }
 
 export interface PickersLayoutProps<TValue, TView extends DateOrTimeView>
diff --git a/packages/x-date-pickers/src/internals/hooks/useDesktopPicker/index.ts b/packages/x-date-pickers/src/internals/hooks/useDesktopPicker/index.ts
index 2d7ec95c5..b7d3c5fd8 100644
--- a/packages/x-date-pickers/src/internals/hooks/useDesktopPicker/index.ts
+++ b/packages/x-date-pickers/src/internals/hooks/useDesktopPicker/index.ts
@@ -2,5 +2,6 @@ export { useDesktopPicker } from './useDesktopPicker';
 export type {
   UseDesktopPickerSlotsComponent,
   UseDesktopPickerSlotsComponentsProps,
+  ExportedUseDesktopPickerSlotsComponentsProps,
   DesktopOnlyPickerProps,
 } from './useDesktopPicker.types';
diff --git a/packages/x-date-pickers/src/internals/hooks/useDesktopPicker/useDesktopPicker.types.ts b/packages/x-date-pickers/src/internals/hooks/useDesktopPicker/useDesktopPicker.types.ts
index 06e81d72d..3c805cd6c 100644
--- a/packages/x-date-pickers/src/internals/hooks/useDesktopPicker/useDesktopPicker.types.ts
+++ b/packages/x-date-pickers/src/internals/hooks/useDesktopPicker/useDesktopPicker.types.ts
@@ -18,6 +18,7 @@ import { BaseFieldProps } from '../../models/fields';
 import {
   ExportedPickersLayoutSlotsComponent,
   ExportedPickersLayoutSlotsComponentsProps,
+  PickersLayoutSlotsComponentsProps,
 } from '../../../PickersLayout/PickersLayout.types';
 import { UsePickerValueNonStaticProps } from '../usePicker/usePickerValue';
 import { UsePickerViewsNonStaticProps, UsePickerViewsProps } from '../usePicker/usePickerViews';
@@ -53,6 +54,10 @@ export interface UseDesktopPickerSlotsComponent<TDate, TView extends DateOrTimeV
 }
 
 export interface UseDesktopPickerSlotsComponentsProps<TDate, TView extends DateOrTimeView>
+  extends ExportedUseDesktopPickerSlotsComponentsProps<TDate, TView>,
+    Pick<PickersLayoutSlotsComponentsProps<TDate | null, TView>, 'toolbar'> {}
+
+export interface ExportedUseDesktopPickerSlotsComponentsProps<TDate, TView extends DateOrTimeView>
   extends PickersPopperSlotsComponentsProps,
     ExportedPickersLayoutSlotsComponentsProps<TDate | null, TView> {
   field?: SlotComponentProps<
diff --git a/packages/x-date-pickers/src/internals/hooks/useMobilePicker/index.ts b/packages/x-date-pickers/src/internals/hooks/useMobilePicker/index.ts
index f38c04003..761b84a6d 100644
--- a/packages/x-date-pickers/src/internals/hooks/useMobilePicker/index.ts
+++ b/packages/x-date-pickers/src/internals/hooks/useMobilePicker/index.ts
@@ -2,5 +2,6 @@ export { useMobilePicker } from './useMobilePicker';
 export type {
   UseMobilePickerSlotsComponent,
   UseMobilePickerSlotsComponentsProps,
+  ExportedUseMobilePickerSlotsComponentsProps,
   MobileOnlyPickerProps,
 } from './useMobilePicker.types';
diff --git a/packages/x-date-pickers/src/internals/hooks/useMobilePicker/useMobilePicker.types.ts b/packages/x-date-pickers/src/internals/hooks/useMobilePicker/useMobilePicker.types.ts
index ceae1cd31..951539bf5 100644
--- a/packages/x-date-pickers/src/internals/hooks/useMobilePicker/useMobilePicker.types.ts
+++ b/packages/x-date-pickers/src/internals/hooks/useMobilePicker/useMobilePicker.types.ts
@@ -16,6 +16,7 @@ import { BaseFieldProps } from '../../models/fields';
 import {
   ExportedPickersLayoutSlotsComponent,
   ExportedPickersLayoutSlotsComponentsProps,
+  PickersLayoutSlotsComponentsProps,
 } from '../../../PickersLayout/PickersLayout.types';
 import { UsePickerValueNonStaticProps } from '../usePicker/usePickerValue';
 import { UsePickerViewsNonStaticProps, UsePickerViewsProps } from '../usePicker/usePickerViews';
@@ -36,7 +37,7 @@ export interface UseMobilePickerSlotsComponent<TDate, TView extends DateOrTimeVi
   TextField?: React.ElementType<TextFieldProps>;
 }
 
-export interface UseMobilePickerSlotsComponentsProps<TDate, TView extends DateOrTimeView>
+export interface ExportedUseMobilePickerSlotsComponentsProps<TDate, TView extends DateOrTimeView>
   extends PickersModalDialogSlotsComponentsProps,
     ExportedPickersLayoutSlotsComponentsProps<TDate | null, TView> {
   field?: SlotComponentProps<
@@ -47,6 +48,10 @@ export interface UseMobilePickerSlotsComponentsProps<TDate, TView extends DateOr
   textField?: SlotComponentProps<typeof TextField, {}, Record<string, any>>;
 }
 
+export interface UseMobilePickerSlotsComponentsProps<TDate, TView extends DateOrTimeView>
+  extends ExportedUseMobilePickerSlotsComponentsProps<TDate, TView>,
+    Pick<PickersLayoutSlotsComponentsProps<TDate | null, TView>, 'toolbar'> {}
+
 export interface MobileOnlyPickerProps<TDate>
   extends BaseNonStaticPickerProps,
     UsePickerValueNonStaticProps<TDate | null>,
```

It is slightly more verbose for sure, but it avoids the `Omit` on each picker, keeping the tweaking at the headless level.

Reply:
Yes, indeed, you understood my requirement correctly.
I've played around with your suggested diff and it seems to cover my requirement with a cleaner solution. I'll go with it, thanks. 👍

Reply:
By the way, I have the same problem with the `Field` slot
The current typing is not correct and fire errors if you try to pass a custom `field` slot on your picker.
We will need to remove it from the exported slot of `useDesktopPicker` and define it at the component level instead (or in a `NonStaticDatePickerSlotsComponent` interface if we want to share between mobile and desktop variants.